[pt] Add incude_last_offset option to EmbeddingBag mean and max

jianyuh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42215 [pt] Add incude_last_offset option to EmbeddingBag mean and max**

Specifically on https://github.com/pytorch/pytorch/pull/27477#discussion_r371402079

We would like to supported with include_last=True overall for other reduction types like mean and max. It now causes further code fragmentation in DPER (https://www.internalfb.com/intern/diff/D22794469/)

Differential Revision: [D22801881](https://our.internmc.facebook.com/intern/diff/D22801881/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22801881/)!

jianyuh committed:
[pt] Add incude_last_offset option to EmbeddingBag mean and max

Specifically on https://github.com/pytorch/pytorch/pull/27477#discussion_r371402079

We would like to supported with include_last=True overall for other reduction types like mean and max. It now causes further code fragmentation in DPER (https://www.internalfb.com/intern/diff/D22794469/)

Differential Revision: [D22801881](https://our.internmc.facebook.com/intern/diff/D22801881/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22801881/)!

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -438,7 +449,13 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> _embedding_bag_cpu_impl(
+
+  at::Tensor bag_size;
+  if (include_last_offset) {
+    bag_size = make_bag_size(offsets.slice(0, 0, offsets.size(0) - 1, 1), indices, mode, requires_grad);
```
this make_bag_size could be optimized to do less temporary tensors, but I guess it's ok for now (maybe add a TODO/comment?)

dzhulgakov commented on test/test_nn.py:
```diff
@@ -10500,19 +10523,22 @@ def test_per_sample_weights_new_offsets(mode, dtype, trainable_scale, include_la
         for dtype, mode, trainable, include_last_offset in itertools.product(dtypes, modes, trainable_scale, include_last_offset):
-            test_per_sample_weights_new_offsets(mode, dtype, trainable, include_last_offset)
+            if mode != 'sum':
+                has_weight = False
```
this actually disables has_weight for future iterations of the loop too :(

just pass `has_weight and mode == 'sum'` below as an arg

dzhulgakov commented on test/test_nn.py:
```diff
@@ -10500,19 +10523,22 @@ def test_per_sample_weights_new_offsets(mode, dtype, trainable_scale, include_la
+        modes = ('sum', 'max', 'mean')
         trainable_scale = (True, False)
         include_last_offset = (True, False)
+        has_weight = (True, False)
```
you probably wanted to include `has_weight` in the for loop iteration too

or you could do for simplicity:

```
modes = (('sum', False), ('sum', True), ('max', False), ('mean', False))

for dtype, (mode, has_weight), trainable, ...
```

dzhulgakov commented on test/test_nn.py:
```diff
@@ -10474,20 +10492,25 @@ def test_per_sample_weights(mode, dtype, trainable_scale):
             offsets = torch.tensor([0, 0, 3, 3, 6], device=device, dtype=torch.long)
 
-            if include_last_offset is True and mode == 'sum':
+            if include_last_offset is True:
```
nit: no need for `is True` :)

dzhulgakov reviewed (approved):
Looking good, just fix the test a bit please! And lint seems to be complaining

houseroad commented on test/test_nn.py:
```diff
@@ -10413,10 +10413,10 @@ def test_EmbeddingBag_per_sample_weights_failures(self, device):
     def _embedding_bag_reference_impl(self, input, weight, offsets=None, mode='sum',
                                       per_sample_weights=None, include_last_offset=False):
-        assert mode == 'sum'
+        assert mode == 'sum' or per_sample_weights == None
```
linter is yelling

jianyuh committed:
Update on "[pt] Add incude_last_offset option to EmbeddingBag mean and max"


Specifically on https://github.com/pytorch/pytorch/pull/27477#discussion_r371402079

We would like to supported with include_last=True overall for other reduction types like mean and max. It now causes further code fragmentation in DPER (https://www.internalfb.com/intern/diff/D22794469/)

Differential Revision: [D22801881](https://our.internmc.facebook.com/intern/diff/D22801881/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22801881/)!

[ghstack-poisoned]

jianyuh committed:
Update on "[pt] Add incude_last_offset option to EmbeddingBag mean and max"


Specifically on https://github.com/pytorch/pytorch/pull/27477#discussion_r371402079

We would like to supported with include_last=True overall for other reduction types like mean and max. It now causes further code fragmentation in DPER (https://www.internalfb.com/intern/diff/D22794469/)

Differential Revision: [D22801881](https://our.internmc.facebook.com/intern/diff/D22801881/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22801881/)!

[ghstack-poisoned]

jianyuh committed:
Update on "[pt] Add incude_last_offset option to EmbeddingBag mean and max"


Specifically on https://github.com/pytorch/pytorch/pull/27477#discussion_r371402079

We would like to supported with include_last=True overall for other reduction types like mean and max. It now causes further code fragmentation in DPER (https://www.internalfb.com/intern/diff/D22794469/)

Differential Revision: [D22801881](https://our.internmc.facebook.com/intern/diff/D22801881/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22801881/)!

[ghstack-poisoned]
